### PR TITLE
TKBrowser and The Authorization Engine for External Authorization Detail

### PR DESCRIPTION
**Pull request by sibinlutoken**

### CODE REVIEW

**Comment by sibinlutoken** [comment]
With this change the sdk user can get rid of the external authorization details and bank authorization. Whenever the external authorization is require, an instance of TKAuthEngine will be created and handles the external authorization details. TKBrowser will be invoked to interact with the users then return the bank authorization. In this pr, when the authorization is required in create transfer token, the auth engine will be used to get the bank auth.  From the developer perspective, create transfer token will always return a token if success.

**Comment by lahosken** [comment]
Alas, I'm still new at ObjC, so I'm slow to figure out this PR 😅 .

I think with this change, when auth is required, some code gets called. If the sdk-using-programmer does nothing, then the sdk shows the user a browser.

Does the sdk-using-programmer have a way to say "instead of using token's browser, use my code instead?" I see interfaces and protocols that seem to suggest this, but I don't see how to make it happen.

(Maybe it can't happen yet but you're planning for the future?)

**Comment by lahosken** [comment]
Sibin came to visit and gave me a walkthrough. tl;dr : So far, the diff doesn't show how an sdk-using-programmer can insert their code; it will eventually be on TokenIO.

File: src/api/TransferTokenBuilder.m [hunk 1]
```diff
@@ -137,10 +133,40 @@ - (void)executeAsync:(OnSuccessWithToken)onSuccess
...
+          authorizedWithExternalAuthorizationDetails:details
+          onSuccess:^(BankAuthorization *auth) {
+              
+              payload.transfer.instructions.source.account.tokenAuthorization.authorization = auth;
+              
+              [[self.member getClient]
+               createTransferToken:payload
+               onSuccess:onSuccess
+               onAuthRequired:^(ExternalAuthorizationDetails *details) {
+                   // Fails to create transfer token with bank authorization
```
[the remark this answers is not in the document]
Comment (sibinlutoken) [reply src/api/TransferTokenBuilder.m]: nice

File: src/authorization/TKAuthorizationEngine.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ */
+@interface TKAuthorizationEngine : NSObject <TKBrowserDelegate>
+/**
+ * Initialize with TKBrowserCreationBlock. If the block is nil or invalid, the engine will
+ * use TKTokenBrowser instead.
+ */
+- (id)initWithBrowserCreationBlock:(TKBrowserCreationBlock)browserCreationBlock;
+
+/**
+ * Authorizes the external authorization details.
```
[the remark this answers is not in the document]
Comment (sibinlutoken) [reply src/authorization/TKAuthorizationEngine.h]: nice

File: src/authorization/TKAuthorizationEngine.m [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    ExternalAuthorizationDetails *details;
+    OnSuccessWithBankAuthorization onSuccess;
+    OnError onError;
+    TKBrowser* browser;
+    
+    BOOL completionUrlIsFound;
+}
+
+/**
+ * The shared pool can prevent TKAuthorizationEngine being recycled by ARC
```
[the remark this answers is not in the document]
Comment (sibinlutoken) [reply src/authorization/TKAuthorizationEngine.m]: update as "The shared pool can prevent TKAuthorizationEngine being recycled **unexpectedly** by ARC"

[the remark this answers is not in the document]
Comment (sibinlutoken) [reply src/authorization/TKAuthorizationEngine.m]: In most case, an async operation will be retain by the operation queue which will prevent the operation object being recycled unexpectedly. Without a mechanism like that, the TKAuthorizationEngine will be recycled by the time method authorizedWithExternalAuthorizationDetails:OnSuccess:OnError ends, while we expect the obj being recycled after it handles onSuccess OnError event. 

The shared pool here take over the job to retain the AuthEngine obj. The AuthEngine will be recycled after revoke function is invoked.

File: src/authorization/TKAuthorizationEngine.m [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+                               } else{
+                                   onError(error);
+                               }
+                           }];
+    return NO;
+}
+
+- (void) authorizationWillCancel:(NSError *)error {
+    onError([NSError errorFromErrorCode:kTKErrorUserCancelled
+                                details:TKLocalizedString(@"User_Cancelled_Authentication", @"User cancelled authentication")
```
[the remark this answers is not in the document]
Comment (sibinlutoken) [reply src/authorization/TKAuthorizationEngine.m]: rename to **authenticationWillCancel**

File: src/authorization/TKBrowser.h [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+ * This method will be invoked when the authorization will cancel by browser.
+ */
+- (void) authorizationWillCancel:(NSError *)error;
+@end
+
+typedef TKBrowser * (^ TKBrowserCreationBlock)(id<TKBrowserDelegate>);
+
+/**
+ * The browser for external authorization. Whenever external authorization is required, the browser will be invoked by TKAuthorizationEngine.
+ * To Use dedicated browsers:
```
[the remark this answers is not in the document]
Comment (sibinlutoken) [reply src/authorization/TKBrowser.h]: dedicated -> customized

File: src/authorization/TKAuthorizationEngine.m [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        // Should not load any request after the completion url is found
+        return NO;
+    }
+    
+    NSString* url = request.URL.absoluteString;
+    
+    NSError *error = nil;
+    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:details.completionPattern
+                                                                           options:nil
+                                                                             error:&error];
```
[the remark this answers is not in the document]
Comment (sibinlutoken) [reply src/authorization/TKAuthorizationEngine.m]: add comment on the header file to explain the error

**Comment by sibinlutoken** [comment]
SDK can use customized browser:
builder.browserCreationBlock = { authEngine in
                return TKTokenAppBrowser(browserDelegate: authEngine)
            }

File: src/api/TransferTokenBuilder.m [hunk 7]
```diff
@@ -137,10 +133,40 @@ - (void)executeAsync:(OnSuccessWithToken)onSuccess
...
+          } onError:^(NSError *error) {
+              onError(error);
+              [authEngine revoke];
+          }];
+     }
+     onError:onError];
 }
 
+- (void)dealloc {
+    NSLog(@"AA");
```
Comment (akalini) [inline src/api/TransferTokenBuilder.m]: Remove?

File: src/authorization/TKAuthorizationEngine.h [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * Authorization Engine handles all the external authorization details in sdk.
+ */
+@interface TKAuthorizationEngine : NSObject <TKBrowserDelegate>
+/**
+ * Initialize with TKBrowserCreationBlock. If the block is nil or invalid, the engine will
+ * use TKTokenBrowser instead.
+ */
+- (id)initWithBrowserCreationBlock:(TKBrowserCreationBlock)browserCreationBlock;
```
Comment (akalini) [inline src/authorization/TKAuthorizationEngine.h]: initWithBrowserFactory?

File: src/authorization/TKAuthorizationEngine.h [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * Authorizes, using the external authorization details.
+ * An regular expression error means the external authorization details object is invalid.
+ * @param details the external authorization details to authorizes
```
Comment (akalini) [inline src/authorization/TKAuthorizationEngine.h]: empty line above.

File: src/authorization/TKAuthorizationEngine.h [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+ * @param onSuccess invoked on success with bank authorization
+ * @param onError invoked on error (Cancellation is a kind of error)
+ */
+- (void)authorizedWithExternalAuthorizationDetails:(ExternalAuthorizationDetails *)details
+                                         onSuccess:(OnSuccessWithBankAuthorization)onSuccess
+                                           onError:(OnError)onError;
+
+/**
+ * Revoking the authorization engine will dismiss the browser. Revoke is necessary after
+ * authorizedWithExternalAuthorizationDetails. This method enhance the flexibility to control the browser.
```
Comment (akalini) [inline src/authorization/TKAuthorizationEngine.h]: long line

File: src/authorization/TKAuthorizationEngine.h [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+ */
+- (void)revoke;
```
Comment (akalini) [inline src/authorization/TKAuthorizationEngine.h]: revoke sounds wrong for the use case. close? or dismiss?

File: src/authorization/TKAuthorizationEngine.h (hunk 10 above)
Comment (akalini) [inline src/authorization/TKAuthorizationEngine.h]: needs to be a link to `authorizedWithExternalAuthorizationDetails`. It's a public API, we need to make sure that the generated docs are nice.

File: src/authorization/TKAuthorizationEngine.h [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+- (id)initWithBrowserCreationBlock:(TKBrowserCreationBlock)browserCreationBlock;
+
+/**
+ * Authorizes, using the external authorization details.
+ * An regular expression error means the external authorization details object is invalid.
+ * @param details the external authorization details to authorizes
+ * @param onSuccess invoked on success with bank authorization
+ * @param onError invoked on error (Cancellation is a kind of error)
+ */
+- (void)authorizedWithExternalAuthorizationDetails:(ExternalAuthorizationDetails *)details
```
Comment (akalini) [inline src/authorization/TKAuthorizationEngine.h]: Should we just call it `authorize`?

File: src/authorization/TKAuthorizationEngine.m [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+    OnError onError;
+    TKBrowser* browser;
+    
+    BOOL completionUrlIsFound;
+}
+
+/**
+ * The shared pool can prevent TKAuthorizationEngine being recycled unexpectedly by ARC
+ */
++ (id)sharedPool {
```
Comment (akalini) [inline src/authorization/TKAuthorizationEngine.m]: Not sure why this is needed. If close or dismiss or revoke is always used to dismiss it, then there is a reference held by the client.

File: src/authorization/TKAuthorizationEngine.m [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+    TKLogDebug(@"Should start loading %@",request)
+    if (completionUrlIsFound) {
+        // Should not load any request after the completion url is found
+        return NO;
+    }
+    
+    NSString* url = request.URL.absoluteString;
+    
+    NSError *error = nil;
+    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:details.completionPattern
```
Comment (akalini) [inline src/authorization/TKAuthorizationEngine.m]: long line

File: src/authorization/TKAuthorizationEngine.m [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+    }
+    
+    completionUrlIsFound = YES;
+    
+    // Download the bank authorization data
+    [NSURLConnection sendAsynchronousRequest:request
+                                       queue:[NSOperationQueue mainQueue]
+                           completionHandler:^(NSURLResponse *response, NSData *data, NSError *error) {
+                               if ( !error )
+                               {
```
Comment (akalini) [inline src/authorization/TKAuthorizationEngine.m]: join with prev line

File: src/authorization/TKAuthorizationEngine.m [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+        return YES;
+    }
+    
+    completionUrlIsFound = YES;
+    
+    // Download the bank authorization data
+    [NSURLConnection sendAsynchronousRequest:request
+                                       queue:[NSOperationQueue mainQueue]
+                           completionHandler:^(NSURLResponse *response, NSData *data, NSError *error) {
+                               if ( !error )
```
Comment (akalini) [inline src/authorization/TKAuthorizationEngine.m]: remove spaces around `!error`

File: src/authorization/TKAuthorizationEngine.m [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+                               {
+                                   NSError *err = nil;
+                                   NSDictionary *dict = [NSJSONSerialization JSONObjectWithData:data
+                                                                                        options:nil
+                                                                                          error:&err];
+                                   if (err != nil) {
+                                       onError(err);
+                                   }
+                                   else {
+                                       BankAuthorization *bankAuth = [TKJson deserializeMessageOfClass:[BankAuthorization class] fromDictionary:dict];
```
Comment (akalini) [inline src/authorization/TKAuthorizationEngine.m]: long line

File: src/authorization/TKAuthorizationEngine.m [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+                                       TKLogDebug(@"Receive BankAuth %@",bankAuth)
```
Comment (akalini) [inline src/authorization/TKAuthorizationEngine.m]: Received

File: src/authorization/TKAuthorizationEngine.m [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+                                       onSuccess(bankAuth);
+                                   }
+                               } else{
```
Comment (akalini) [inline src/authorization/TKAuthorizationEngine.m]: space before {

File: src/authorization/TKAuthorizationEngine.m [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+                                   onError(error);
+                               }
+                           }];
+    return NO;
+}
+
+- (void) authenticationWillCancel:(NSError *)error {
+    onError([NSError errorFromErrorCode:kTKErrorUserCancelled
```
Comment (akalini) [inline src/authorization/TKAuthorizationEngine.m]: long lines

File: src/authorization/TKAuthorizationEngine.m [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+#pragma mark - private methods
+- (TKBrowser *)_createBrowser {
+    if (browserCreationBlock != nil) {
+        TKBrowser *browser_ = browserCreationBlock(self);
+        if (browser_ != nil) {
+            return browser_;
+        }
+    }
+    // Use TKTokenBrowser instead
+    return [[TKTokenBrowser alloc] initWithBrowserDelegate:self];
```
Comment (akalini) [inline src/authorization/TKAuthorizationEngine.m]: This decision should be done up the stack, in the TokenIO?

File: src/authorization/TKBrowser.h [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+
+@class TKBrowser;
+
+/**
+ * TKAuthorizationEngine observes TKBrowser with this protocol
+ */
+@protocol TKBrowserDelegate
+@required
+/**
+ * Return Yes if the browser shall load this request; Otherwise the browser will discard the request.
```
Comment (akalini) [inline src/authorization/TKBrowser.h]: I don't understand what this method accomplishes. When would you want to discard the request?

File: src/authorization/TKBrowser.h [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+ */
+- (BOOL) shouldStartLoadWithRequest:(NSURLRequest *)request;
+/**
+ * This method will be invoked when the authorization will cancel by browser.
+ */
+- (void) authenticationWillCancel:(NSError *)error;
```
Comment (akalini) [inline src/authorization/TKBrowser.h]: This has nothing to do with authentication, right? It's just a browser that you are popping up, right?

File: src/authorization/TKTokenBrowser.m [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+            [self.delegate authenticationWillCancel:nil];
+        }
+        
+        [rootViewController presentViewController:browserViewController
+                                         animated:true
+                                       completion:^{
+                                           [browserViewController loadUrl:url];
+                                       }];
+    }
+    else {
```
Comment (akalini) [inline src/authorization/TKTokenBrowser.m]: join with prev line

File: src/authorization/TKTokenBrowser.m [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+
+- (void)webView:(UIWebView *)webView didFailLoadWithError:(NSError *)error {
+    if (error.code >= 400) {
+        // Error code 102 is the result of not passing redirect URL through so we can ignore that
+        [self.delegate authenticationWillCancel:error];
+        [browserViewController dismissViewControllerAnimated:true completion:nil];
+    }
+}
+
+- (BOOL)webView:(UIWebView *)webView shouldStartLoadWithRequest:(NSURLRequest *)request navigationType:(UIWebViewNavigationType)navigationType{
```
Comment (akalini) [inline src/authorization/TKTokenBrowser.m]: long line

File: src/authorization/TKTokenBrowserViewController.m [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+
+- (BOOL)webView:(UIWebView *)webView
+shouldStartLoadWithRequest:(NSURLRequest *)request
+ navigationType:(UIWebViewNavigationType)navigationType {
+    return [delegate webView:webView shouldStartLoadWithRequest:request navigationType:navigationType];
+}
+
+- (void)webViewDidFinishLoad:(UIWebView *)webView {
+    // TODO: Remove this js after the problem fix on bank-demo
+    NSString *jsRemoveYobleeTopRightCloseButton = @"var element = document.getElementsByClassName('yodlee-font-icon svg_close close-modal-window closeIcon right hide-for-mobile-only')[0];  element.style.display = 'none';";
```
Comment (akalini) [inline src/authorization/TKTokenBrowserViewController.m]: long line

File: src/authorization/TKTokenBrowserViewController.xib [hunk 27]
```diff
@@ -0,0 +1,100 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (akalini) [inline src/authorization/TKTokenBrowserViewController.xib]: Can we move all the UI related stuff into a separate package / directory?

File: src/util/TKError.m [hunk 28]
```diff
@@ -17,10 +17,17 @@ + (instancetype)errorFromErrorCode:(TKErrorCode)errorCode details:(NSString *)de
...
-                               code:errorCode
-                           userInfo:@{ NSLocalizedDescriptionKey: details ,
-                                       TKEncapsulatedErrorKey: error }];
+    if (error != nil) {
+        return [NSError errorWithDomain:kTokenErrorDomain
+                                   code:errorCode
+                               userInfo:@{ NSLocalizedDescriptionKey: details ,
+                                           TKEncapsulatedErrorKey: error }];
+    }
+    else {
```
Comment (akalini) [inline src/util/TKError.m]: Join with prev line.

File: src/util/TKError.m [hunk 29]
```diff
@@ -17,10 +17,17 @@ + (instancetype)errorFromErrorCode:(TKErrorCode)errorCode details:(NSString *)de
 }
 
 + (instancetype)errorFromErrorCode:(TKErrorCode)errorCode details:(NSString *)details encapsulatedError:(NSError *)error {
```
Comment (akalini) [inline src/util/TKError.m]: Long line

File: src/api/TransferTokenBuilder.m [hunk 30]
```diff
@@ -137,10 +133,40 @@ - (void)executeAsync:(OnSuccessWithToken)onSuccess
...
+         [authEngine
+          authorizedWithExternalAuthorizationDetails:details
+          onSuccess:^(BankAuthorization *auth) {
+              
+              payload.transfer.instructions.source.account.tokenAuthorization.authorization = auth;
+              
+              [[self.member getClient]
+               createTransferToken:payload
+               onSuccess:onSuccess
+               onAuthRequired:^(ExternalAuthorizationDetails *details) {
```
[the remark this answers is not in the document]
Comment (sibinlutoken) [reply src/api/TransferTokenBuilder.m]: fixed.

**Pull request merged**